File: home/dot_local/share/lazyvim/capabilities/foundation.mjs
import { captureCommandOutput } from "../lib/commands.mjs";
import { defineCapability } from "./contract.mjs";

function expectPrefix(actual, expected, label) {
  if (!actual.startsWith(expected)) {
    throw new Error(`${label}: expected ${expected}, got ${actual}`);
  }
}

export default defineCapability({
  id: "foundation",
  verify({ platform, versions }) {
    expectPrefix(
      captureCommandOutput("rg", ["--version"]),
      `ripgrep ${versions.ripgrep}`,
      "ripgrep",
    );
    expectPrefix(
      captureCommandOutput("fd", ["--version"]),
      `fd ${versions.fdMajor}.`,
      "fd",
    );
    expectPrefix(
      captureCommandOutput("fzf", ["--version"]),
      versions.fzf,
      "fzf",
    );
    if (
      !captureCommandOutput("lazygit", ["--version"]).includes(
        `version=${versions.lazygit}`,
      )
    ) {
      throw new Error("Unexpected lazygit version");
    }
    if (
      !captureCommandOutput("tree-sitter", ["--version"]).includes(
        versions.treeSitter,
      )
    ) {
      throw new Error("Unexpected tree-sitter version");
    }
    const rainfrog = platform.managedToolExecutable("rainfrog");
    if (!(platform.platformName === "win32" && process.arch === "arm64")) {
      captureCommandOutput(rainfrog, ["--version"]);
    }
  },
});

File: home/dot_local/share/lazyvim/lua/lazyvim_capabilities/capabilities/foundation.lua
local commands = require("lazyvim_capabilities.commands")
local define = require("lazyvim_capabilities.contract")

local function prefix(actual, expected, label)
	assert(vim.startswith(actual, expected), ("%s: expected %s, got %s"):format(label, expected, actual))
end

return define({
	id = "foundation",
	verify = function(context)
		local v = context.versions
		prefix(commands.capture("rg", { "--version" }), "ripgrep " .. v.ripgrep, "ripgrep")
		prefix(commands.capture("fd", { "--version" }), "fd " .. v.fd_major .. ".", "fd")
		prefix(commands.capture("fzf", { "--version" }), v.fzf, "fzf")
		assert(
			commands.capture("lazygit", { "--version" }):find("version=" .. v.lazygit, 1, true),
			"Unexpected lazygit version"
		)
		assert(
			commands.capture("tree-sitter", { "--version" }):find(v.tree_sitter, 1, true),
			"Unexpected tree-sitter version"
		)
		if not (context.platform.name == "win32" and jit.arch == "arm64") then
			commands.capture(context.platform.tool("rainfrog"), { "--version" })
		end
	end,
})
